File: csharp/Behavioral/20-Strategy/Strategy/EmergencyRateStrategy.cs
namespace Strategy;

// Wraps another strategy and adds an after-hours surcharge on top
public class EmergencyRateStrategy : IPricingStrategy
{
    private readonly IPricingStrategy _baseStrategy;
    private readonly decimal _surchargeRate;

    public EmergencyRateStrategy(
        IPricingStrategy baseStrategy,
        decimal surchargeRate = 0.5m)
    {
        _baseStrategy  = baseStrategy;
        _surchargeRate = surchargeRate;
    }

    public string Name => $"Emergency ({_surchargeRate:P0} surcharge)";

    public decimal Calculate(JobCard job) =>
        _baseStrategy.Calculate(job) * (1 + _surchargeRate);

    public string Explain(JobCard job)
    {
        var baseAmount = _baseStrategy.Calculate(job);
        var surcharge  = baseAmount * _surchargeRate;

        return $"{_baseStrategy.Explain(job)} " +
               $"After-hours surcharge of {_surchargeRate:P0} " +
               $"(N${surcharge:N2}) applied.";
    }
}

File: typescript/Behavioral/20-Strategy/Strategy/EmergencyRateStrategy.ts
import { IPricingStrategy } from "./IPricingStrategy";
import type { JobCard } from "../Model/JobCard";

// Wraps another strategy and adds an after-hours surcharge on top
export class EmergencyRateStrategy implements IPricingStrategy {
  constructor(
    private readonly baseStrategy: IPricingStrategy,
    private readonly surchargeRate: number = 0.5,
  ) {}

  get name(): string {
    return `Emergency (${(this.surchargeRate * 100).toFixed(0)}% surcharge)`;
  }

  calculate(job: JobCard): number {
    return this.baseStrategy.calculate(job) * (1 + this.surchargeRate);
  }

  explain(job: JobCard): string {
    const baseAmount = this.baseStrategy.calculate(job);
    const surcharge = baseAmount * this.surchargeRate;

    return (
      `${this.baseStrategy.explain(job)} ` +
      `After-hours surcharge of ${(this.surchargeRate * 100).toFixed(0)}% ` +
      `(N$${surcharge.toFixed(2)}) applied.`
    );
  }
}
